### PR TITLE
fix: avoid redirect previous established conn for ambient mode

### PR DESCRIPTION
**Pull request by alpha-baby**
**Please provide a description of this PR:**

#48876

### CODE REVIEW

**Review by bleggett — COMMENTED** [review]
Couple of things

- This has to be optional - not everyone would want to silently allow existing unsecured traffic to continue to flow even after Istio says it's secured.

- Whether optional or not, it needs an enforcable deadline - how long should we let an unsecured, invisible-to-policy connection live? Forever? 5 minutes?

I don't think this can be done purely in iptables, for the above reasons.

**Comment by alpha-baby** [comment]
> Couple of things
> 
> * This has to be optional, and IMO off-by-default - not everyone would want to silently allow existing unsecured traffic to continue to flow even after Istio says workload traffic is secured.
> * Whether optional or not, unsecured, existing connections need an enforceable deadline - how long should we let an unsecured, invisible-to-policy connection live? Forever? 5 minutes?
> 
> I don't think this can be done purely in iptables, for the above reasons.

In a corporate environment, any changes that could potentially impact critical services indeed require very careful planning and execution. Migrating large-scale services to a service mesh architecture like Istio, particularly when it involves substantial changes to the network layer, should be done in the most graceful and smooth manner possible for the users.

We can imagine a scenario where a user's company operates an extremely important online service with substantial traffic, running in a particular namespace, such as with thousands of machines. If this namespace were to be incorporated into Istio, there could be a moment where the network connections of these thousands of machines would be forcibly severed, rendering them unable to provide service. I believe that such an "accident" would be intolerable for any company. However, there may be some previously insecure connections that have been running this way before, with minimal risk. I think that compared to security, the normal operation of the service is more important.

That being said, if the user is concerned about these insecure connections and needs to disconnect them, I believe the best solution would be for the user to perform a rolling restart of their programs. This approach would offer users the highest degree of control and prevent the occurrence of "online accidents."

**Comment by zengyuxing007** [comment]
> > Couple of things
> > 
> > * This has to be optional, and IMO off-by-default - not everyone would want to silently allow existing unsecured traffic to continue to flow even after Istio says workload traffic is secured.
> > * Whether optional or not, unsecured, existing connections need an enforceable deadline - how long should we let an unsecured, invisible-to-policy connection live? Forever? 5 minutes?
> > 
> > I don't think this can be done purely in iptables, for the above reasons.
> 
> In a corporate environment, any changes that could potentially impact critical services indeed require very careful planning and execution. Migrating large-scale services to a service mesh architecture like Istio, particularly when it involves substantial changes to the network layer, should be done in the most graceful and smooth manner possible for the users.
> 
> We can imagine a scenario where a user's company operates an extremely important online service with substantial traffic, running in a particular namespace, such as with thousands of machines. If this namespace were to be incorporated into Istio, there could be a moment where the network connections of these thousands of machines would be forcibly severed, rendering them unable to provide service. I believe that such an "accident" would be intolerable for any company. However, there may be some previously insecure connections that have been running this way before, with minimal risk. I think that compared to security, the normal operation of the service is more important.
> 
> That being said, if the user is concerned about these insecure connections and needs to disconnect them, I believe the best solution would be for the user to perform a rolling restart of their programs. This approach would offer users the highest degree of control and prevent the occurrence of "online accidents."

I also think it's the need to have a switch to make a decision on whether to keep an existing connection or not.

This switch is multidimensional and the user can control the granularity. Consider adding related options under MeshConfig and ProxyConfig

>> existing connections need an enforceable deadline - how long should we let an unsecured, invisible-to-policy connection live? Forever? 5 minutes?

IMO, this probably be left off for now, and we can make a decision on whether or not to add it at a later stage based on user feedback; in most cases, existing non-secure connections will be closed with the next release of the workload

**Comment by alpha-baby** [comment]
The most suitable solution at present is to add a startup argument to the install-cni program, which will determine whether there is a need to forcibly disconnect the previous connections. Based on the above discussion, the default value of not forcibly disconnecting the previous connections is a better choice. Do you have any other better suggestions?

new flag: `--reset-previous-connections`, default value: `false`

**Comment by bleggett** [comment]
> In a corporate environment, any changes that could potentially impact critical services indeed require very careful planning and execution. Migrating large-scale services to a service mesh architecture like Istio, particularly when it involves substantial changes to the network layer, should be done in the most graceful and smooth manner possible for the users.

Fully agree, that's why we recommend node cordons and blue/green node pools, which give operators total control over disruptions.

> We can imagine a scenario where a user's company operates an extremely important online service with substantial traffic, running in a particular namespace, such as with thousands of machines. If this namespace were to be incorporated into Istio, there could be a moment where the network connections of these thousands of machines would be forcibly severed, rendering them unable to provide service. I believe that such an "accident" would be intolerable for any company. However, there may be some previously insecure connections that have been running this way before, with minimal risk. I think that compared to security, the normal operation of the service is more important.

Yes, and operating clusters at this scale, with or without Istio, requires careful planning - kubernetes upgrades, node upgrades, or hardware failure happen routinely and would also break these applications, which is why this kind of fragility must be solved primarily with operational strategies entirely outside the scope of Istio. If deployed applications cannot recover from brief network interruptions, they will be chronically unstable in any K8S environment, with or without Istio.

> That being said, if the user is concerned about these insecure connections and needs to disconnect them, I believe the best solution would be for the user to perform a rolling restart of their programs. This approach would offer users the highest degree of control and prevent the occurrence of "online accidents."

If they can do that, and are willing to do that, do they need this?

**Comment by alpha-baby** [comment]
> Yes, and operating clusters at this scale, with or without Istio, requires careful planning - kubernetes upgrades, node upgrades, or hardware failure happen routinely and would also break these applications, which is why this kind of fragility must be solved primarily with operational strategies entirely outside the scope of Istio. If deployed applications have no redundancy and cannot recover from brief network interruptions, they will be chronically unstable in any K8S environment, with or without Istio.

Typically, we deploy a microservice within the same namespace, and when we incorporate that namespace into Istio, we inject redirect rules into the entire service application at once, rather than gradually as a canary deployment. This is one of the reasons that I believe could lead to "online accidents." I acknowledge the point you made about various unexpected events that could cause stability issues in our system; however, those are accidental and unavoidable. But adding our system to Istio is not an accidental event.

Although I think that avoiding "online accidents" is more important than "network security," I can't guarantee that everyone will have the same requirements as I do. We can continue to wait for feedback from other users, but from what we currently see, it seems necessary to add such a switch to control whether to disconnect previous connections.

**Comment by louiscryan** [comment]
Its not clear to me that such an option actually provides meaningful 'safety'. It has the potential effect of creating a delay between the operational intent (enable ambient) and the outcome (traffic works or breaks). Any time we create gaps in time like these between intent and outcome we make root causing a failure much harder.

If enabling ztunnel is going to break an application you would rather know that immediately and have the cause be clear.  Since connections can stay open for the entire lifetime of an application, weeks or months in some cases, we will have user reports complaining about why their applications are broken by a rollout they did a month ago.

If there is concern that the _rate_ of connection termination is higher than an individual application could tolerate (seems odd) then an argument could be made for controlling how quickly connections were churned.

It seems like the root of the proposal is fear about the scope of change, i.e enabling ambient for a whole namespace at once. I think controls over the scope are a much more useful area to explore. Examples here include:

- Apply ambient to workloads with label=canary  (probably the best practice, but requires setup by operators)
- Apply ambient to workloads with start time > X.  (likely OK but not as good as above, useful in combination with restarts)
- Apply ambient to a maximum of 1 pod in any deployment. (easy for operators but more complex in Istio)

All of the above are in combination with the ability to control rollout scope at the Deployment/Pod & Namespace level as is provided today.

**Review by bleggett — CHANGES_REQUESTED** [review]
> Typically, we deploy a microservice within the same namespace, and when we incorporate that namespace into Istio, we inject redirect rules into the entire service application at once, rather than gradually as a canary deployment. This is one of the reasons that I believe could lead to "online accidents." I acknowledge the point you made about various unexpected events that could cause stability issues in our system; however, those are accidental and unavoidable. But adding our system to Istio is not an accidental event.

> Although I think that avoiding "online accidents" is more important than "network security," I can't guarantee that everyone will have the same requirements as I do. We can continue to wait for feedback from other users, but from what we currently see, it seems necessary to add such a switch to control whether to disconnect previous connections.

I agree, but what you are talking about is a blast radius problem - "I want to upgrade workloads selectively in a namespace, and I do not feel comfortable upgrading everything in the namespace at once."

That is where I am in agreement with @louiscryan - if we want or need blast radius controls for operators that are more granular than "the whole namespace" - let's do _that_ instead. Letting unsecured connections for fragile apps continue for $undefined-length-of-time doesn't solve the problem, creates new ones for us, and actually makes things harder and less obvious for operators at scale.

> Its not clear to me that such an option actually provides meaningful 'safety'. It has the potential effect of creating a delay between the operational intent (enable ambient) and the outcome (traffic works or breaks). Any time we create gaps in time like these between intent and outcome we make root causing a failure much harder.

:+1: on this.

- If this does not have a user-defined hard timeout on when connections will be dropped, I don't think it's safe or useful to merge.

- If the need is better blast radius controls around enabling ambient, we should do _that_ instead of opaque connection-level allowlisting.

**Comment by alpha-baby** [comment]
@zengyuxing007 @louiscryan thx for your comment.

Apologies, I may have diverged too much in this PR with many inappropriate examples, causing our discussion to exceed the scope of the problem that this PR is meant to address.

The core issue that this PR is intended to raise is:

Do we need to mandatorily terminate business application TCP connections?

From our current discussions, I see two perspectives:

We need to immediately enforce the termination of connections, because we want to ensure that all connections are secure.
We should not enforce termination previous TCP connections but rather let users decide for themselves. 


I believe we should not enforce the termination of business application TCP connections. Why can't we enforce termination? First of all, faced with large and complex business systems, Istio developers and maintainers simply do not understand these systems or the internal details of business applications well enough. We have no idea what business applications are doing over these TCP connections and what data is being sent. For example, a TCP connection could be executing a MySQL transaction, or it could be transferring files. In such scenarios, I believe it is not appropriate to forcefully terminate the connection. 

Take the company I work for, for instance; we are primarily in the financial industry, which has very strict security requirements. Our online systems already use mTLS, and we have the capability to encrypt data. Therefore, we do not need to utilize Istio's mTLS encryption capabilities. 

Previously, other istio's developers thought that terminating previous TCP connections would have a minor impact because applications have retry logic. However, I want to point out that, for Istio developers, it’s not necessary to understand so many complex business systems in detail, nor should we assume that business applications have retry logic. For example, in our company, an online microservices system node may have tens of thousands of connections. While our RPC framework has retry capabilities, re-establishing tens of thousands of connections would also incur significant costs. 

In my earlier statements, I mentioned that adding a namespace to Istio could pose stability risks to the entire system within that namespace. You suggested using a canary strategy to address this, which I think is excellent. I believe Istio needs this capability as well, but this is not the problem that this PR aims to solve. Why does it lead to stability risks? The fundamental reason is the forced termination of already established TCP connections. That is the core issue this PR seeks to resolve. 

Forcing TCP connection termination has different advantages and disadvantages, and the changes I have made so far allow users to choose, rather than absolutely prohibiting the termination of previous TCP connections.

**Comment by alpha-baby** [comment]
@bleggett @louiscryan @zengyuxing007

**Comment by bleggett** [comment]
> @zengyuxing007 @louiscryan thx for your comment.
> 
> Apologies, I may have diverged too much in this PR with many inappropriate examples, causing our discussion to exceed the scope of the problem that this PR is meant to address.
> 
> The core issue that this PR is intended to raise is:
> 
> Do we need to mandatorily terminate business application TCP connections?

If they are installing Istio for the first time and enrolling workloads into ambient, we will, as a mandatory step, disrupt existing workload connections, yes.

This is because Istio exists to provide telemetry and policy enforcement around workload connections, and installing it is an explicit act of intent - if we do not disrupt existing workload connections on initial install, we necessarily violate the intent.

This isn't unique to ambient, this is how Istio has always worked - it's how all service meshes work. It's how CNIs all work.

> From our current discussions, I see two perspectives:
> 
> We need to immediately enforce the termination of connections, because we want to ensure that all connections are secure. We should not enforce termination previous TCP connections but rather let users decide for themselves.
> 
> I believe we should not enforce the termination of business application TCP connections. Why can't we enforce termination? First of all, faced with large and complex business systems, Istio developers and maintainers simply do not understand these systems or the internal details of business applications well enough. We have no idea what business applications are doing over these TCP connections and what data is being sent. For example, a TCP connection could be executing a MySQL transaction, or it could be transferring files. In such scenarios, I believe it is not appropriate to forcefully terminate the connection.
> 
> Take the company I work for, for instance; we are primarily in the financial industry, which has very strict security requirements. Our online systems already use mTLS, and we have the capability to encrypt data. Therefore, we do not need to utilize Istio's mTLS encryption capabilities.

I certainly understand that, but Istio's mTLS encryption guarantees are a fundamental aspect of the product, and not something we plan to make optional.

> Previously, other istio's developers thought that terminating previous TCP connections would have a minor impact because applications have retry logic. However, I want to point out that, for Istio developers, it’s not necessary to understand so many complex business systems in detail, nor should we assume that business applications have retry logic. For example, in our company, an online microservices system node may have tens of thousands of connections. While our RPC framework has retry capabilities, re-establishing tens of thousands of connections would also incur significant costs.
> 
> In my earlier statements, I mentioned that adding a namespace to Istio could pose stability risks to the entire system within that namespace. You suggested using a canary strategy to address this, which I think is excellent. I believe Istio needs this capability as well, but this is not the problem that this PR aims to solve. Why does it lead to stability risks? The fundamental reason is the forced termination of already established TCP connections. That is the core issue this PR seeks to resolve.

As already discussed - Istio will, and must, terminate workload connections on initial install.

What we still _do_ need to solve for is ensuring that upgrades from one functional version of Istio to another functional version of Istio do not require terminating workload connections a second time - I am in total agreement with you that in the real world, in large deployments, non-disruptive upgrades are a requirement. We do not yet have a story for this in OSS ambient, and it is not currently in scope for reaching beta status, which is the immediate work target for ambient.

**Comment by alpha-baby** [comment]
@zengyuxing007 
> I certainly understand that, but Istio's mTLS encryption guarantees are a fundamental aspect of the product, and not something we plan to make optional.
> As already discussed - Istio will, and must, terminate workload connections on initial install.
What we still do need to solve for is ensuring that upgrades from one functional version of Istio to another functional version of Istio do not require terminating workload connections a second time - I am in total agreement with you that in the real world, in large deployments, non-disruptive upgrades are a requirement. We do not yet have a story for this in OSS ambient, and it is not currently in scope for reaching beta status, which is the immediate work target for ambient. I personally would support a design for this (post-beta, at least).

Istio's mTLS encryption guarantees are indeed an important feature of the product. However, if using this feature requires forcibly terminating existing TCP connections (which could disrupt service requests), this is unacceptable to our team (we place a high value on the stability of our production systems, and all functional priorities are secondary). Therefore, in this PR, I've simply added a switch that allows me or other users to choose whether or not to forcibly terminate existing TCP connections. Moreover, the default behavior of this switch is consistent with previous behaviors.

In our company, the system is vast, and stability is the primary factor we need to consider, as any risk that arises could lead to significant economic losses. I believe this is the case for any large corporation.

File: releasenotes/notes/49106.yaml
Comment (linsun) [inline releasenotes/notes/49106.yaml:9]: default value is true, i assume, worth mention it here.

**Comment by linsun** [comment]
Thinking loud here (without reading details of your PR other than UX side), is it possible to leverage permissive mode where in permissive mode we continue to allow previous established conn to continue, without introducing a new flag (ambient-reset-previous-connections) as you proposed?

**Comment by alpha-baby** [comment]
@linsun 
> Thinking loud here (without reading details of your PR other than UX side), is it possible to leverage permissive mode where in permissive mode we continue to allow previous established conn to continue, without introducing a new flag (ambient-reset-previous-connections) as you proposed?

sorry, what is permissive mode? I can't understand. 

Are you suggesting renaming the `--ambient-reset-previous-connections` flag?

**Comment by alpha-baby** [comment]
@linsun permissive mode is https://istio.io/latest/docs/reference/config/security/peer_authentication/#PeerAuthentication-MutualTLS-Mode ?

read val(`permissive mode`) is difficult in `istio-cni`

**Comment by kdorosh** [comment]
My two cents;

How would your workloads handle connection termination in the event of hardware failure? Or node upgrade? These are things that already have to be dealt with, regardless of the complex systems above istio.

So long as the rate of connection disruption is limited, I think it is reasonable to let istio forcibly terminate connections. Applications should be able to handle this; the strongest argument I have heard against this is that it may be an unnecessary cost.. but how expensive is it? ($$?)

**Comment by zengyuxing007** [comment]
> My two cents;
> 
> How would your workloads handle connection termination in the event of hardware failure? Or node upgrade? These are things that already have to be dealt with, regardless of the complex systems above istio.
> 
> So long as the rate of connection disruption is limited, I think it is reasonable to let istio forcibly terminate connections. Applications should be able to handle this; if they can't then the systems already have larger architectural problems
> 
> I would consider this flag an anti pattern, but with some level of business pressure perhaps making sense as a stopgap. For this situation, I think it would be better to fork istio than commit an anti pattern upstream.

Don't agree with this, we can't assume that all Applications are suitable for terminate connection behavior. Like PeerAuthentication, some customers have mTLS requirements, and some have DISABLE requirements.

**Comment by alpha-baby** [comment]
> My two cents;
> 
> How would your workloads handle connection termination in the event of hardware failure? Or node upgrade? These are things that already have to be dealt with, regardless of the complex systems above istio.

The Istio development team and the business application development team are two distinct groups. Whether business application developers make their programs robust is their own concern. However, as the Istio development team, we should consider the impact of forcibly terminating connections on business applications. The example you cited refers to hardware failures, which are unavoidable, but forced disconnections when using or upgrading Istio can be avoided. This is not an accidental situation, which is the biggest difference from hardware failures.

> So long as the rate of connection disruption is limited, I think it is reasonable to let istio forcibly terminate connections. Applications should be able to handle this; if they can't then the systems already have larger architectural problems

Controlling the frequency of disconnections is also quite complex. As pointed out in previous discussions, connections should not be terminated in parts, as this could delay faults to an unpredictable time. Business application development teams are responsible for maintaining the robustness of their systems, but not all teams may have this capability. Our company is very large, with many business application development teams, and we cannot communicate with each one to require them to be robust.

**Comment by howardjohn** [comment]
If they are not suitable then you can also just not enable ambient on running pods and instead enable it only for new pods. There is already a per-pod annotation to enroll to ambient

**Comment by alpha-baby** [comment]
> If they are not suitable then you can also just not enable ambient on running pods and instead enable it only for new pods. There is already a per-pod annotation to enroll to ambient

Following what you've suggested, we would only enable the ambient mesh for newly created pods. However, with the massive number of pods in our company, requiring business applications to redeploy would be very costly. It would be more efficient to dynamically inject iptables redirection rules during the pod's runtime, which would spare the Istio development team from having to communicate extensively with business application development teams and from forcing them to restart their applications (although some business application development teams might want to forcibly close previous TCP connections, most in our company would prefer not to).

The business application development teams also wouldn't need to worry that enabling Istio's ambient mode would impact the stability of their programs, which would make the internal promotion and adoption of Istio by the development team much faster and easier.

**Comment by bleggett** [comment]
> > If they are not suitable then you can also just not enable ambient on running pods and instead enable it only for new pods. There is already a per-pod annotation to enroll to ambient
> 
> Following what you've suggested, we would only enable the ambient mesh for newly created pods. However, with the massive number of pods in our company, requiring business applications to redeploy would be very costly. It would be more efficient to dynamically inject iptables redirection rules during the pod's runtime, which would spare the Istio development team from having to communicate extensively with business application development teams and from forcing them to restart their applications.

They should not have to forcibly restart their applications when enabling ambient. They should only have to do this if any connection drop of any kind in-cluster required them to restart their applications - if this is a problem, then this is an operational problem those applications will have in a Kubernetes context with or without Istio.

If the applications are that fragile, they are highly stateful workloads, and either should be run as StatefulSets in K8S and/or managed separately - whether you use Istio or not.

> 
> The business application development teams also wouldn't need to worry that enabling Istio's ambient mode would impact the stability of their programs, which would make the internal promotion and adoption of Istio by the development team much faster and easier.

How do those teams absorb things like node upgrades today?

**Comment by alpha-baby** [comment]
> How do those teams absorb things like node upgrades or brief networking outages today? They certainly have to worry about those, and those definitely impact the stability of their programs, correct?
We cannot break the Istio security model and policy enforcement for 99% of workloads, just to support the 1% of things running as K8S StatefulSets, which anyway in K8S must be carefully managed on their own even without Istio, as a basic operational principle.

Is the security capability of Istio that important? I believe it is very important, but it also depends on the user's own scenario. For example, in Istio, the [PeerAuthentication-MutualTLS-Mode](https://istio.io/latest/docs/reference/config/security/peer_authentication/#PeerAuthentication-MutualTLS-Mode) is also left for the users to choose, isn't it?

Whether to forcibly close previous TCP connections is a question we can set aside for now to see more feedback from other Istio users. It's similar to the idea of providing a switch in Istio for users to turn off mTLS if they choose to.

**Comment by hzxuzhonghu** [comment]
Is this still valid with in pod redirection now?

**Comment by alpha-baby** [comment]
> Is this still valid with in pod redirection now?

In the case of in-pod redirection with dynamically injected iptables rules, would you prefer not to forcibly close the connections?

**Comment by hzxuzhonghu** [comment]
With inpod redirection, i think this case can not occur. Redirection rules is setup before application is running

**Comment by alpha-baby** [comment]
> With inpod redirection, i think this case can not occur. Redirection rules is setup before application is running

The scenario you are referring to is likely one where redirection rules are injected before the pod starts. However, the scenario I'm describing involves pods that are already running, and when a particular namespace is dynamically added to Istio, iptables rules are injected on the fly during pod execution.

This dynamic capability is very beneficial as it allows Istio developers to quickly promote and integrate Istio within their company without the need for application services to restart pods, unlike the sidecar pattern.

**Comment by hzxuzhonghu** [comment]
I donot think it should be so complex, and in pod redirection does not support such scenario too

**Comment by alpha-baby** [comment]
> I donot think it should be so complex, and in pod redirection does not support such scenario too

Currently, if a namespace is dynamically added to Istio's management, redirection rules are injected dynamically. 

You might think that dynamically injecting redirection rules could be risky, but as long as it doesn't affect existing TCP connections, I believe the risk is manageable. 

Of course, the Istio community could choose to remove this dynamic capability.

**Comment by bleggett** [comment]
> > With inpod redirection, i think this case can not occur. Redirection rules is setup before application is running
> 
> The scenario you are referring to is likely one where redirection rules are injected before the pod starts. However, the scenario I'm describing involves pods that are already running, and when a particular namespace is dynamically added to Istio, iptables rules are injected on the fly during pod execution.

If adding an entire namespace at once is too disruptive, operators are today able to add workloads individually with labels, rather than the entire namespace at once, so this scenario can easily be mitigated in 2 different ways in the current code.

1. By not annotating the entire namespace, and opting in workloads within that namespace at the operator's desired cadence with workload labels.

2. By annotating fragile workloads you do not want disrupted with `ambient.istio.io/redirection: disabled`, then annotating the namespace. These workloads will not be redirected or injected or disrupted.


Both of these methods already offer the operator total control at the workload level, so the operator already has full control over what happens when. 

Opting in a single workload to ambient capture - which is explicit operator intent - and then having that explicit opt-in _not_ take effect for the workload for some variable and undefined length of time isn't a feature Istio intends to support at this time.

**Comment by alpha-baby** [comment]
> > > With inpod redirection, i think this case can not occur. Redirection rules is setup before application is running
> > 
> > 
> > The scenario you are referring to is likely one where redirection rules are injected before the pod starts. However, the scenario I'm describing involves pods that are already running, and when a particular namespace is dynamically added to Istio, iptables rules are injected on the fly during pod execution.
> 
> If adding an entire namespace at once is too disruptive, operators are today able to add workloads individually with labels, rather than the entire namespace at once, so this scenario can easily be mitigated in 2 different ways in the current code.
> 
> 1. By not labelling the entire namespace, and opting in workloads within that namespace at the operator's desired cadence with workload-level labels.
> 2. By annotating fragile workloads you do not want disrupted with `ambient.istio.io/redirection: disabled`, then labelling the entire namespace. These excluded workloads will not be redirected or injected or disrupted.
> 
> Either of these methods already offer the operator total control at the workload level, so the operator already has full control over what happens when.
> 
> Opting in a single workload to ambient capture - which is explicit operator intent - and then having that explicit opt-in _not_ take effect for the workload for some variable and undefined length of time isn't a feature Istio intends to support at this time.

Being able to dynamically inject iptables rules during the pod's runtime is extremely valuable as it enables rapid onboarding of business application developers to Istio. I believe this capability will be particularly useful among commercial companies in China.

The Istio ambient mode enables dynamic injection of iptables rules into pods, which is uncontested, right?
I acknowledge that the methods you described can indeed mitigate the impact brought by dynamic injection of iptables rules. From the perspective of a namespace, there seems to be no issue.
However, from the perspective of a pod, if iptables rules are injected during the pod's runtime, it would forcibly terminate TCP connections. When would this happen? It would occur only when TCP packets are in flow and a reset packet is received.
At least from our company's standpoint, we cannot accept TCP connections being forcibly terminated. Of course, I can also opt for the ability to inject iptables rules only and only before the pod starts. This would then require the business pods to be restarted.



@zengyuxing007 @hzxuzhonghu

**Comment by bleggett** [comment]
>  At least from our company's standpoint, we cannot accept TCP connections being forcibly terminated.

If you have pods that are this fragile, and do not have the ability to tolerate even brief network instability to the point where they can never be restarted, then you have pods that even Kubernetes would struggle to properly manage, service mesh aside. You are describing node-critical pods that cannot tolerate any network interruption or K8S rescheduling.

My advice would be to exclude these specific pods from any capture with the `ambient.istio.io/redirection: disabled` annotation I mentioned, and wait to include them for capture by removing the annotation until you would have to restart them anyway, such as during a K8S node upgrade.

**Comment by alpha-baby** [comment]
> > At least from our company's standpoint, we cannot accept TCP connections being forcibly terminated.
> 
> If you have pods that are this fragile, and do not have the ability to tolerate even brief network instability to the point where they can never be restarted, then you have pods that even Kubernetes would struggle to properly manage, service mesh aside. You are describing node-critical pods that cannot tolerate any network interruption or K8S rescheduling.
> 
> My advice would be to exclude these specific pods from any capture with the `ambient.istio.io/redirection: disabled` annotation I mentioned, and wait to include them for capture by removing the annotation until you would have to restart them anyway, such as during a K8S node upgrade.

Your suggestion is quite good and may be suitable for many people, but it's not the best solution in our company's scenario. The best solution for us is as presented in my PR, where the iptables rule would not disrupt the existing TCP connections already established by the business applications, and would only affect new incremental TCP connections.

**Pull request closed without merging**
